core/utils/pool_handler.hpp

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright Quadrivium LLC
+ * All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <boost/asio/io_context.hpp>
+
+namespace kagome {
+
+  inline bool runningInThisThread(
+      std::shared_ptr<boost::asio::io_context> ioc) {
+    return ioc->get_executor().running_in_this_thread();
+  }
+
+  class PoolHandler {
+   public:
+    PoolHandler(PoolHandler &&) = delete;
+    PoolHandler(const PoolHandler &) = delete;
+
+    PoolHandler &operator=(PoolHandler &&) = delete;
+    PoolHandler &operator=(const PoolHandler &) = delete;
+
+    // Next nested struct and deleted ctor added to avoid unintended injections
+    struct Inject {
+      explicit Inject() = default;
+    };
+    explicit PoolHandler(Inject, ...);
+
+    explicit PoolHandler(std::shared_ptr<boost::asio::io_context> io_context)
+        : is_active_{false}, ioc_{std::move(io_context)} {}
+    ~PoolHandler() = default;
```

Comment:
Must be constructed as "active",
or will lose callbacks before `AppStateManager` calls `start()`
```suggestion
        : is_active_{true}, ioc_{std::move(io_context)} {}
```

Reply:
Making it active is responsibility of using class. To avoid callback loosing, class-owner should start pool handler at 'prepare' phase.

Reply:
Should we fix `PoolHandler { is_active_ = true }` constructor,
or each class constructor `Foo(PoolHandler pool_handler) { pool_handler.start() }`?